pkg/arvo/app/chat-hook.hoon

```diff
@@ -0,0 +1,364 @@
+::
++$  card
+  $%  [%diff [%chat-update chat-update]]
+      [%quit ~]
+      [%poke wire dock poke]
+      [%pull wire dock ~]
+      [%peer wire dock path]
+  ==
+::
++$  state
+  $%  [%0 state-zero]
+  ==
+::
++$  state-zero
+  $:  synced=(map path ship)
+      boned=(map wire (list bone))
+  ==
+::
++$  poke
+  $%  [%chat-action chat-action]
+      [%permission-action permission-action]
+  ==
+::
+--
+::
+|_  [bol=bowl:gall state]
+::
+++  this  .
+::
+++  prep
+  |=  old=(unit state)
+  ^-  (quip move _this)
+  ?~  old
+    :_  this
+    [ost.bol %peer /permissions [our.bol %permission-store] /updates]~
+  [~ this(+<+ u.old)]
+::
+++  poke-noun
+  |=  a=*
+  ^-  (quip move _this)
+  ~&  synced
+  ~&  boned
+  [~ this]
+::
+++  poke-json
+  |=  jon=json
+  ^-  (quip move _this)
+  (poke-chat-action (json-to-action jon))
+::
+++  poke-chat-action
+  |=  act=chat-action
+  ^-  (quip move _this)
+  ?>  ?=(%message -.act)
+  ::  local
+  :_  this
+  ?:  (team:title our.bol src.bol)
+    ?.  (~(has by synced) path.act)
+      ~
+    =/  ship  (~(got by synced) path.act)
+    =/  appl  ?:(=(ship our.bol) %chat-store %chat-hook)
+    [ost.bol %poke / [ship appl] [%chat-action act]]~
+  ::  foreign
+  =/  ship  (~(get by synced) path.act)
+  ?~  ship
+    ~
+  ?.  =(u.ship our.bol)
+    ~
+  ::  scry permissions to check if write is permitted
+  ?.  (permitted-scry [(scot %p src.bol) %chat (weld path.act /write)])
+    ~
+  =.  author.envelope.act  src.bol
+  =.  when.envelope.act  now.bol
+  [ost.bol %poke / [our.bol %chat-store] [%chat-action act]]~
+::
+++  poke-chat-hook-action
+  |=  act=chat-hook-action
+  ^-  (quip move _this)
+  ?-  -.act
+      %add-owned
+    ?>  (team:title our.bol src.bol)
+    =/  chat-path  [%mailbox path.act]
+    ?:  (~(has by synced) path.act)
+      [~ this]
+    =.  synced  (~(put by synced) path.act our.bol)
+    :_  (track-bone chat-path)
+    %+  weld
+      [ost.bol %peer chat-path [our.bol %chat-store] chat-path]~
+    (create-permission [%chat path.act] security.act)
+  ::
+      %add-synced
+    ?>  (team:title our.bol src.bol)
+    =/  chat-path  [%mailbox (scot %p ship.act) path.act]
+    ?:  (~(has by synced) [(scot %p ship.act) path.act])
+      [~ this]
+    =.  synced  (~(put by synced) [(scot %p ship.act) path.act] ship.act)
+    :_  (track-bone chat-path)
+    [ost.bol %peer chat-path [ship.act %chat-hook] chat-path]~
+  ::
+      %remove
+    =/  ship  (~(get by synced) path.act)
+    ?~  ship
+      [~ this]
+    ?:  &(=(u.ship our.bol) (team:title our.bol src.bol))
+      ::  delete one of our.bol own paths
+      :_
+      %_  this
+        synced  (~(del by synced) path.act)
+        boned  (~(del by boned) [%mailbox path.act])
+      ==
+      %-  zing
+      :~  (pull-wire [%mailbox path.act])
+          (delete-permission [%chat path.act])
+          ^-  (list move)
+          %+  turn  (prey:pubsub:userlib [%mailbox path.act] bol)
+          |=  [=bone *]
+          ^-  move
```

Comment:
You shouldn't need both this `^-  move` and the `^-  (list move)` -- either should be enough for both documentary purposes and the type system.

Reply:
Good point, thanks.

Reply:
Removed cast in latest commit.